janusgraph-couchbase/src/main/java/org/janusgraph/diskstorage/couchbase/CouchbaseKeyColumnValueStore.java

```diff
@@ -0,0 +1,438 @@
+/*
+ * Copyright 2023 Couchbase, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.janusgraph.diskstorage.couchbase;
+
+import com.couchbase.client.core.error.CouchbaseException;
+import com.couchbase.client.java.Bucket;
+import com.couchbase.client.java.Cluster;
+import com.couchbase.client.java.Collection;
+import com.couchbase.client.java.json.JsonArray;
+import com.couchbase.client.java.json.JsonObject;
+import com.couchbase.client.java.manager.collection.CollectionManager;
+import com.couchbase.client.java.manager.collection.CollectionSpec;
+import com.couchbase.client.java.manager.collection.ScopeSpec;
+import com.couchbase.client.java.query.QueryOptions;
+import com.couchbase.client.java.query.QueryResult;
+import com.couchbase.client.java.query.QueryScanConsistency;
+import com.google.common.collect.Iterators;
+import org.janusgraph.diskstorage.BackendException;
+import org.janusgraph.diskstorage.Entry;
+import org.janusgraph.diskstorage.EntryList;
+import org.janusgraph.diskstorage.EntryMetaData;
+import org.janusgraph.diskstorage.PermanentBackendException;
+import org.janusgraph.diskstorage.StaticBuffer;
+import org.janusgraph.diskstorage.TemporaryBackendException;
+import org.janusgraph.diskstorage.keycolumnvalue.KCVMutation;
+import org.janusgraph.diskstorage.keycolumnvalue.KeyColumnValueStore;
+import org.janusgraph.diskstorage.keycolumnvalue.KeyIterator;
+import org.janusgraph.diskstorage.keycolumnvalue.KeyRangeQuery;
+import org.janusgraph.diskstorage.keycolumnvalue.KeySliceQuery;
+import org.janusgraph.diskstorage.keycolumnvalue.KeySlicesIterator;
+import org.janusgraph.diskstorage.keycolumnvalue.MultiSlicesQuery;
+import org.janusgraph.diskstorage.keycolumnvalue.SliceQuery;
+import org.janusgraph.diskstorage.keycolumnvalue.StoreTransaction;
+import org.janusgraph.diskstorage.util.RecordIterator;
+import org.janusgraph.diskstorage.util.StaticArrayEntry;
+import org.janusgraph.diskstorage.util.StaticArrayEntryList;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.couchbase.client.java.query.QueryOptions.queryOptions;
+
+public class CouchbaseKeyColumnValueStore implements KeyColumnValueStore {
```

Comment:
In JanusGraph 1.0.0 a new method `getMultiSlices` is added into `KeyColumnValueStore`. This method may receive multiple slice queries which a multi-query needs to execute. For example, multiple edges or multiple properties should be fetched for a set of vertices. I.e. `g.V(v1,v2,v3).outE("edge1", "edge2", "edge3")`.
The default implementation is not optimized and it uses blocking calls. I.e. first it will retrieve "edge1" edges for v1,v2,v3, then after those edges were retrieved it will fetch "edge2", and later after edge2 is retrieved it will fetch "edge3". 
Instead, for performance consideration it's recommended to overwrite `getMultiSlices` method and fetch them in parallel (if possible).
An example can be found in CQL storage implementation where multiple slice queries are fetched in parallel using asynchronous capabilities of the DataStax Java Driver:
https://github.com/JanusGraph/janusgraph/blob/09d49b276f667e28333797a7e9db719209dcf5a5/janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java#L400

If you think Couchbase is able to process slice queries in parallel then it might be better to overwrite `getMultiSlices` method. However, it's not required.

Reply:
Will look into this tomorrow. Thank you for reviewing and suggesting some improvements!